neo/Consensus/ConsensusService.cs

```diff
@@ -225,18 +273,273 @@ private void OnPersistCompleted(Block block)
+                            // a backup and accepting our own previous request.
+                            context.State |= ConsensusState.Backup;
+                            // Since our Primary flag is still set, we must act as though we already sent the request.
+                            context.State |= ConsensusState.RequestSent;
+
+                            ChangeTimer(TimeSpan.FromSeconds(Blockchain.SecondsPerBlock << (context.ViewNumber + 1)));
+                        }
+                        OnPrepareRequestReceived(prepareRequestPayload, prepareRequest);
+                        preparationHash = prepareRequestPayload.Hash;
+                    }
+                    else
+                    {
+                        // Can't use anything from the recovery message if we are primary and haven't sent a prepare
+                        // request and no prepare request was present in the recovery message.
+                        if (myIndexIsPrimary) return;
+
+                        // If we have no `Preparation` hash we can't regenerate any of the `PrepareRequest` messages.
+                        if (message.PreparationHash == null) return;
+                        preparationHash = message.PreparationHash;
+                    }
+                }
+
+                for (ushort i = 0; i < context.Validators.Length; i++)
+                {
+                    // If we are missing this preparation.
+                    if (context.PreparationPayloads[i] != null) continue;
+                    if (i == context.PrimaryIndex) continue;
+                    // If the recovery message has this preparations
+                    if (!message.PreparationMessages.ContainsKey(i)) continue;
+                    var prepareResponseMsg = new PrepareResponse {PreparationHash = preparationHash};
+                    prepareResponseMsg.ViewNumber = context.ViewNumber;
+                    var regeneratedPrepareResponse = ((ConsensusContext) context).RegenerateSignedPayload(
+                        prepareResponseMsg, i, message.PreparationMessages[i].InvocationScript);
+                    if (regeneratedPrepareResponse.Verify(snap) &&
+                        PerformBasicConsensusPayloadPreChecks(regeneratedPrepareResponse))
+                        OnPrepareResponseReceived(regeneratedPrepareResponse, prepareResponseMsg);
+                }
+            }
+
+            if (!context.State.HasFlag(ConsensusState.CommitSent)) return;
+
+            RestoreCommits(message);
+        }
+
+        private void OnRecoveryMessageReceived(ConsensusPayload payload, RecoveryMessage message)
+        {
+            if (context.BlockIndex > payload.BlockIndex) return;
+            if (context.State.HasFlag(ConsensusState.BlockSent)) return;
+            Snapshot snap =  Blockchain.Singleton.GetSnapshot();
+            if (payload.BlockIndex > snap.Height + 1) return;
```

Comment:
Is this necessary? We have already checked it in `PerformBasicConsensusPayloadPreChecks()`.

https://github.com/neo-project/neo/blob/86d40c67a4a7abaf89e7c5209b1ca6d2bbd54c4f/neo/Consensus/ConsensusService.cs#L229

Reply:
You are correct they are all redundant. When I originally wrote them, I think I was thinking the code did not go through the `OnConsensusPayload()` path for some reason.

Reply:
I removed all these redundant checks.